### PR TITLE
Profile page

### PR DESCRIPTION
**Pull request by Dzybaty**

### CODE REVIEW

File: src/reducers/user.js [hunk 1]
```diff
@@ -1,16 +1,12 @@
...
     id: '',
-    email: '',
-    login: '',
 };
 
 const userReducer = (state = defaultState, action) => {
     switch (action.type) {
         case 'SET_USER':
             return Object.assign({}, state, {
                 id: action.payload.id,
```
Comment (vviital) [inline src/reducers/user.js]: Why we need to set only user id?

File: src/sagas/index.js [hunk 2]
```diff
@@ -1,48 +1,57 @@
...
 }
 
 function* getToken(action) {
     const { payload } = action;
     try {
-        const res = yield call(axios.post.bind(axios), 'http://127.0.0.1:3000/tokens', payload);
-        yield call(setLocalStorage, {token: res.data.token});
-        yield call(getUser);
+        const res = yield call(api.getToken, payload);
+        ls.setItem('token', res);
```
Comment (vviital) [inline src/sagas/index.js]: Please use `call` effect.

File: src/sagas/index.js [hunk 3]
```diff
@@ -1,48 +1,57 @@
...
     yield call(clearUser);
 }
 
-function setLocalStorage(payload) {
-    ls.setItem('token', payload.token);
+function* getUser() {
+    let user = {
+        id: '',
+    };
+    if (ls.getItem('token')) {
```
Comment (vviital) [inline src/sagas/index.js]: The same.

File: src/sagas/index.js [hunk 4]
```diff
@@ -1,48 +1,57 @@
...
+       user = jwt(ls.getItem('token'));
```
Comment (vviital) [inline src/sagas/index.js]: The same for `getItem`

File: src/reducers/user.js (hunk 1 above)
Comment (Dzybaty) [reply src/reducers/user.js]: We are checking only the presence of 'id' field. Any other user information is stored in 'user' prop.

File: src/api/API.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import Client from './client';
+
+const ls = window.localStorage;
+
+class API {
+    async getToken(payload) {
+        const res = await Client.post('http://127.0.0.1:3000/tokens', payload);
```
Comment (vviital) [inline src/api/API.js]: Please make this prefix configurable in the constructor:
```javascript
buildURL(path) {
   return `${this.base}/${path}`
}
...
async getToken(payload) {
  const res = await Client.post(this.buildURL('tokens'), payload);
  ...
}
```

File: src/api/client.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const axios = require('axios');
+
+class Client {
+    static post(url ,  payload, options) {
```
Comment (vviital) [inline src/api/client.js]: Please add method / methodWithAuth for all needed HTTP verbs:
- GET
- POST
- PUT
- PATCH
- DELETE

File: src/api/client.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        return axios.post(url, payload, options);
+    }
+
+    static get(url, options) {
+        return axios.get(url, options);
+    }
+
+    static getWithAuth(url, options) {
```
Comment (vviital) [inline src/api/client.js]:
```javascript
static getWithAuth(url, options = {})
```

File: src/api/client.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        return axios.get(url, {
```
Comment (vviital) [inline src/api/client.js]:
```javascript
const token = ls.get('token');
const headers = Object.assign({}, _.get(options, 'headers', {}), { Authorization: `Bearer ${token}` });
return axios.get(url, { ...options, headers });
```

File: src/api/API.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    async getProfile(id) {
+        const token = ls.getItem('token');
+        const res = await Client.getWithAuth(`http://127.0.0.1:3000/profiles/${id}`, {token: token});
+        return res.data;
+    }
+
+    async getProfiles() {
+        const token = ls.getItem('token');
```
Comment (vviital) [inline src/api/API.js]: We do not need token at this level.

File: src/api/API.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        const res = await Client.post('http://127.0.0.1:3000/tokens', payload);
+        if (res.data && res.data === 401)
+        {
+            return undefined;
+        }
+        return res.data.token;
+    }
+
+    async getProfile(id) {
+        const token = ls.getItem('token');
```
Comment (vviital) [inline src/api/API.js]: We do not need token at this level.

File: src/api/API.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import Client from './client';
+
+const ls = window.localStorage;
+
+class API {
+    async getToken(payload) {
+        const res = await Client.post('http://127.0.0.1:3000/tokens', payload);
+        if (res.data && res.data === 401)
+        {
+            return undefined;
```
Comment (vviital) [inline src/api/API.js]: let's return empty string.

File: src/api/client.js (hunk 8 above)
Comment (Dzybaty) [reply src/api/client.js]: Not sure if I fully understand this line:
`const headers = Object.assign({}, _.get(options, 'headers', {}), { Authorization: `Bearer ${token}` });`

File: src/sagas/index.js [hunk 12]
```diff
@@ -1,48 +1,65 @@
...
+        type: 'SET_PROFILES',
+        payload: res,
+    });
+}
+
 function* clearUser() {
     yield put({
-        type: "SET_USER", payload: {}
+        type: "SET_USER",
+        payload: {
```
Comment (vviital) [inline src/sagas/index.js]: I believe you can add `CLEAR_USER` action and set default state in the reducer.

**Comment by vviital** [comment]
LGTM!

**Pull request merged**